Update README with dependency notes (#456)
ava-labs/awm-relayer: README.md

    - Full-service cross-chain message delivery application that is configurable to listen to specific source and destination chain pairs and relay messages according to its configured rules.
2. [Signature Aggregator](signature-aggregator/README.md)
    - Lightweight API that requests and aggregates signatures from validators for any ICM message, and returns a valid signed message that the user can then self-deliver to the intended destination chain.

> ## Updating dependencies and E2E testing

Author: This isn't about updating E2E tests but a section on "Updating dependencies" and "E2E testing"

> Applications in this repository depend on the following upstream repositories, both directly in terms of code imports defined in `go.mod` file as well as indirectly for E2E tests where binary versions are used to spin up the test network via `tmpnet`:

Reviewer: ```suggestion
Applications in this repository depend on the following upstream repositories, both directly in terms of code imports defined in the `go.mod` file as well as indirectly for E2E tests where binary versions are used to spin up the test network via `tmpnet`:
```

1.  [avalanchego](https://github.com/ava-labs/avalanchego/)
2.  [coreth](https://github.com/ava-labs/coreth) (indirectly)
3.  [subnet-evm](https://github.com/ava-labs/subnet-evm)

> [!NOTE]

> > We require any commits referenced in our `main` branch to be present in the default branches of the repositories above, but during active development it might be useful to work against changes in progress that are still on feature branches.

Reviewer 2: How do we enforce this?

Author: Code review for now? I assume we could write a github action that does it as well but doesn't seem like high priority.

Reviewer 3: if this was not done correctly, would the e2e tests build correctly?

Author: Yeah they would. E2E tests just need the commits to be reachable on the remote. They don't care if it's in a feature or default branch

> When developing such features that require updates to one or more of the above, care must be taken to understand where the relevant code comes from. The binaries of applications built in this repo are built against versions referenced in the `go.mod` file. The E2E tests run against a simulated network running locally that is started by calling a separately compiled `avalanchego` binary as well as its plugins. These are compiled based on the values of `AVALANCHEGO_VERSION` in the local checkout of `subnet-evm` when running the tests locally and directly in this repository's `./scripts/versions.sh` when running E2E tests remotely through GitHub actions. 

Reviewer 2: This is currently only in a branch, but we equate the versions in `go.mod` and the binary installed for the e2e tests like so in `teleporter`: https://github.com/ava-labs/teleporter/blob/staking-contract/scripts/versions.sh#L32

We could simplify this documentation by incorporating this approach here.

Author: Yeah but it's possible and valid to do both. At the very least this explains which code comes from where.

Author: I.e. it might be useful to test a version of relayer code against a tmpnet running a separate version of the code. I agree that we should make the `extract-commit` change in this repo as well, but for development and troubleshooting I still want some version of the documentation above.

> `avalanchego` and `coreth` have a direct circular dependency and this repository is only indirectly dependent on `coreth` but directly on `avalanchego`. Therefore if any updates are required from the `coreth` side, a corresponding `avalanchego` commit referencing those changes is required. On the other hand `subnet-evm` just depends directly on `avalanchego`.

Reviewer 2: ```suggestion
`avalanchego` and `coreth` have a direct circular dependency and this repository is only indirectly dependent on `coreth` but directly on `avalanchego`. Therefore if any updates are required from the `coreth` side, a corresponding `avalanchego` commit referencing those changes is required. On the other hand `subnet-evm` just depends directly on `avalanchego`.
```
```suggestion
`avalanchego` and `coreth` have a direct circular dependency and this repository is only indirectly dependent on `coreth` but directly dependent on `avalanchego`. Therefore if any updates are required from the `coreth` side, a corresponding `avalanchego` commit referencing those changes is required. On the other hand `subnet-evm` just depends directly on `avalanchego`.
```

> The most complicated example case that can arise above is that a feature depends on a new change in `coreth`. In this case, the developer needs to:

Reviewer 2: Looks like this section is incomplete.

Author: The cases below this actually explain the workflow that I'm introducing here. I can make them bold instead of headings to make it clearer perhaps?

Reviewer 3: I think it's just the wording, usually when using : i expect a list to follow right after

Reviewer 3: "in this case the developer would need to perform the following steps for either remote or local e2e testing cases." or similar

Reviewer 2: Yeah, the separation between headings is confusing. Perhaps explicitly mention the headings, or mention the "following sections"

### Remote E2E testing through GitHub Actions

> 1. If an `avalanchego` commit referencing this change in it's `go.mod` file doesn't exist yet it needs to be added.

Reviewer: ```suggestion
1. If an `avalanchego` commit referencing this change in its `go.mod` file doesn't exist yet then it needs to be added.
```

> 2. In `subnet-evm` add a commit that references the `avalanchego` commit from above in both its `go.mod` file as well as `scripts/versions.sh` file. 

Reviewer: ```suggestion
2. Add a commit in `subnet-evm` that references the `avalanchego` commit from above in both its `go.mod` file as well as its `scripts/versions.sh` file. 
```

> 3. Create a new commit in this repository referencing `avalanchego` and `subnet-evm` directly and `coreth` indirectly as well as update references in `scripts/version.sh` file for both `AVALANCHEGO_VERSION` and `SUBNET_EVM_VERSION`.

Reviewer: ```suggestion
3. Create a new commit in this repository referencing `avalanchego` and `subnet-evm` directly and `coreth` indirectly as well as update references in the `scripts/version.sh` file for both `AVALANCHEGO_VERSION` and `SUBNET_EVM_VERSION`.
```

>  If all of the commits mentioned above are published to GitHub branches, this will enable running of E2E tests through the CI.

Reviewer: Perhaps change this sentence to: "Publishing all of the commits mentioned above to GithHub branches will enable running of the E2E tests through the CI."

### Local E2E testing

 Local testing through running `./scripts/e2e_test.sh` also requires local checkout of the `subnet-evm` repository with the commit matching the `go.mod` and `SUBNET_EVM_VERSION`.

>  If they do not match, tests might still run but fail with unexpected errors. 

Reviewer 2: Not necessary to merge this, but we should look into removing the local subnet-evm requirement for awm-relayer, to match teleporter

Author: Yeah I considered this originally but I do find this useful to be able to work against purely local versions of `subnet-evm` without needing to pollute the remote with branches for local testing. We could make it optional and if not present pull subnet-evm based on `go.mod`/`versions.sh`

Reviewer 2: That would be useful, though we should be consistent between repos. We should also support that more generically for all upstreams, including avalanchego and coreth.
